[FIX] account: check duplicate accounts by normalized code

filisbits opened the pull request:
**Description of the issue/feature this PR addresses:**
Function `_pre_reload_data` checks for the existance of accounts with the same code to avoid creating a duplicate. It does so however comparing with the code before normalizing it with the length from the template, failing to find possible duplicates with a normalized code.

**Current behavior before PR:**
If accounts with non-normalized codes _(eg 172)_ are being loaded, there's a risk that an account with a matching normalized code _(eg 172000)_ already exists in the database. If the corresponding xmlid is not pointing at it, another account with the same code will be created triggering the ValidationError.

**Desired behavior after PR is merged:**
The comparison with existing accounts is made with the normalized code avoiding this conflicts.



---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

filisbits commented:
I found this issue while working on an Upgrade script. Let me know if this approach makes sense.

lordkrandel commented:
In `l10n_fr_fec_import` before importing the file we search for accounts that start with the same prefix, excluding righthanded zeroes. --> 123, 1230, 12300, 123000, 1230000...

Maybe we're installing a CoA with 6 digits and there already are accounts with 8 for any reason.
I would ask the opinion of @william-andre as an expert of this piece of code.

filisbits commented:
The issue happens when [l10n_be_reports is loaded](https://github.com/odoo/enterprise/blob/17.0/l10n_be_reports/__init__.py). It looks like some 3 digits codes are loaded while already corresponding normalized codes exist. It could also be that the upgrade script that I mentioned above is not the right approach.

william-andre commented on addons/account/tests/test_chart_template.py:
```diff
@@ -355,17 +360,27 @@ def local_get_data(self, template_code):
 
+        # create an account with a normalized existing code
+        problematic_account_b = self.env['account.account'].create({
+            'code': '123000',
```
But we wouldn't find the account with `123` without the `000` anymore.

filisbits replied:
Indeed, I was focused on the duplicates and I hadn't considered that case.

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -295,9 +296,10 @@ def tax_template_changed(tax, template):
+                    if not account or (account and account.code != normalized_code):
                         existing_account = self.env['account.account'].search([
-                            ('code', '=', values['code']),
+                            ('code', '=', normalized_code),
```
Maybe something like this?
```suggestion
                            ('code', '=like', values['code'] + "%"),
```

The best would be with a regex...
The easiest way (but not super elegant) would be to use `_search` with `add_where`.

```python
query = self.env['account.account']._search([('company_id', '=', company.id)])
query.add_where("account_account.code SIMILAR TO '%s0*'", [values['code']])
existing_account = self.env['account.account'].browse(query)
```

filisbits replied:
I've made the changes. It's not very elegant... but it seems to get the job done.

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -295,18 +295,19 @@ def tax_template_changed(tax, template):
-                            ('company_id', '=', company.id),
-                        ])
-                        if existing_account:
+                    normalized_code = f'{values["code"]:<0{int(template_data.get("code_digits", 6))}}'
+                    if not account or account.code != normalized_code:
```
```suggestion
                    if not account or not re.match(f'^{values["code"]0*$', account.code):
```

filisbits replied:
Done.

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -295,18 +295,19 @@ def tax_template_changed(tax, template):
+                        query = self.env['account.account']._search([('company_id', '=', company.id)])
+                        query.add_where("account_account.code SIMILAR TO %s", [f'{values["code"]}0*'])
+                        accounts = self.env['account.account'].browse(query)
+                        account = accounts.sorted(key=lambda x: x.code != normalized_code)[0] if accounts else None
```
Why are you making the sort here? Can't you just take the first one? (not even sure it is valid to have `123` and `123000` for instance :thinking: (even if there is no hard constraint)

filisbits replied:
I thought it more appropriate to give priority the one with the completely matching code if it comes to that. You never know what you might find :man_shrugging:

william-andre reviewed (approved):
Thanks!

@robodoo r+

william-andre commented:
@filisbits I just noticed that the test had been removed, why?

filisbits commented:
> @filisbits I just noticed that the test had been removed, why?

With the changes, the case I added was basically the same as the test that already existed so I removed it. I could have modified it for a case with code including zeroes but I preferred to keep it simple.
